### ISSUE TITLE
`RET503`: Conflict with `--warn-unreachable` in `mypy`

### ISSUE THREAD

**Issue by ngnpope** [charliermarsh/ruff#2602]
I strongly suspect this one is in the realms of "cannot do without typing support", but it's at least worth having in the issue tracker for documentation purposes. Maybe it's possible to special case, but we'll never capture all possibilities.

Consider the following:
```python
import pytest
import typing

@pytest.fixture
def test_function(request: pytest.FixtureRequest) -> typing.Any:
    if request.param == "something":
        return request.getfixturevalue("whatever")
    pytest.fail("Invalid parameters.")
```

I get the following output:
```console
❯ ruff --version
ruff 0.0.241

❯ mypy --version
mypy 0.991 (compiled: yes)

❯ ruff --isolated --select RET503 bug.py 
bug.py:8:5: RET503 [*] Missing explicit `return` at the end of function able to return non-`None` value
Found 1 error.
[*] 1 potentially fixable with the --fix option.

❯ mypy --strict --warn-unreachable bug.py 
Success: no issues found in 1 source file

❯ ruff --isolated --select RET503 --fix bug.py 
Found 1 error (1 fixed, 0 remaining).

❯ mypy --strict --warn-unreachable bug.py 
bug.py:9: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)
```

The issue here is that `pytest.fail()` is typed to return `typing.NoReturn` as it always raises.

**Comment by charliermarsh** [comment]
We could consider treating `pytest.fail` equivalently to raising an exception, which I think would avoid `RET503`.

**Comment by ngnpope** [comment]
Also `pytest.exit()`, `pytest.skip()`, and `pytest.xfail()` in addition to `pytest.fail()`.

**Comment by charliermarsh** [comment]
I have to look back at the code, but I think it's reasonably well-setup to do that...

**Comment by ngnpope** [comment]
Came across another case of this: [`Field.fail()`](https://github.com/typeddjango/djangorestframework-stubs/blob/bf9d3837eaf0b1e2928f28bad48c1f617813376b/rest_framework-stubs/fields.pyi#L121) in Django REST Framework.

This one is probably a bit harder though as it'd be called using `self.fail()` when overriding one of the methods in a subclass... Without typing support to identify `typing.NoReturn`, it's a bit tricky.

**Comment by charliermarsh** [comment]
Much improved by #2701 and friends.
